src/Database/Redis/URL.hs

```diff
@@ -0,0 +1,56 @@
+module Database.Redis.URL
+    ( parseConnectInfo
+    ) where
+
+import Control.Error.Util (note)
+import Control.Monad (guard)
+import Control.Monad.Plus (partial)
+import Data.Maybe (fromMaybe)
+import Database.Redis.Core (ConnectInfo(..), defaultConnectInfo)
+import Database.Redis.ProtocolPipelining (PortID(..))
+import Network.HTTP.Base
+import Network.URI (parseURI, uriPath, uriScheme)
+import Text.Read (readMaybe)
+
+import qualified Data.ByteString.Char8 as C8
+
+-- | Parse a @'ConnectInfo'@ from a URL
+--
+-- Username is ignored, path is used to specify the database:
+--
+-- >>> parseConnectInfo "redis://username:password@host:42/2"
+-- Right (ConnInfo {connectHost = "host", connectPort = PortNumber 42, connectAuth = Just "password", connectDatabase = 2, connectMaxConnections = 50, connectMaxIdleTime = 30s, connectTimeout = Nothing})
+--
+-- N.B. invalid (non-integer) databases are ignored and revert to default:
+--
+-- >>> connectDatabase <$> parseConnectInfo "redis://username:password@host:42/db"
+-- Right 0
+--
+-- The scheme is validated, to prevent mixing up configurations:
+--
+-- >>> parseConnectInfo "postgres://"
+-- Left "Wrong scheme"
+--
+-- Beyond that, all values are optional. Omitted values are taken from
+-- @'defaultConnectInfo'@:
+--
+-- >>> parseConnectInfo "redis://"
+-- Right (ConnInfo {connectHost = "localhost", connectPort = PortNumber 6379, connectAuth = Nothing, connectDatabase = 0, connectMaxConnections = 50, connectMaxIdleTime = 30s, connectTimeout = Nothing})
+--
+parseConnectInfo :: String -> Either String ConnectInfo
+parseConnectInfo url = do
+    uri <- note "Invalid URI" $ parseURI url
+    note "Wrong scheme" $ guard $ uriScheme uri == "redis:"
+    uriAuth <- note "Missing or invalid Authority"
+        $ parseURIAuthority
+        $ uriToAuthorityString uri
+
+    return defaultConnectInfo
+        { connectHost = fromMaybe (connectHost defaultConnectInfo)
+            $ partial (not . null) $ host uriAuth
+        , connectPort = maybe (connectPort defaultConnectInfo)
+            (PortNumber . fromIntegral) $ port uriAuth
+        , connectAuth = C8.pack <$> password uriAuth
+        , connectDatabase = fromMaybe (connectDatabase defaultConnectInfo)
+            $ readMaybe =<< partial (not . null) (dropWhile (== '/') $ uriPath uri)
```

Comment:
I think it's inconsistent with the rest of the values to fall back to a default if the `read` fails. I would prefer we explicitly read the db number and error if it's not a number, this clearly indicates that you have some configuration problems or mis-understood the format which is expected. It'd also explicitly state that you accept an empty string as a valid case. E.g.:

```
let dbNumPart = dropWhile (== '/') (uriPath uri)
let eDbNum = 
      if null dbNumPart 
      then Right (connectDatabase defaultConnectInfo) 
      else maybe (Left "Error reading db number: " <> dbNumPart) Right (readMaybe dbNumPart)
```